File: 02/solution.rb
require_relative "../utils"
include Utils

def part1(input)
  input
    .map { |a| a.split.map(&:to_i).sort }
    .map { |a| a[-1] - a[0] }
    .reduce(:+)
end

def part2(input)
  input
    .map { |row|
      row
        .split
        .map(&:to_i)
        .sort
        .combination(2)
        .find { |smaller, bigger| bigger % smaller == 0 }
    }
    .map { |smaller, bigger| bigger / smaller }
    .reduce(:+)
end

input = read_input("input")
test_input = read_input("test_input")

check(18, part1(test_input))
puts "Part 1: #{part1(input)}"

check(9, part2(test_input))
puts "Part 2: #{part2(input)}"

File: 02/solution.dart
import '../utils.dart';

dynamic part1(List<String> input) {
  var doubles = 0;
  var triples = 0;
  for (var line in input) {
    var chars = line.codeUnits;
    var counts = <int, int>{};
    for (var char in chars) {
      counts[char] = (counts[char] ?? 0) + 1;
    }
    if (counts.values.contains(2)) {
      doubles++;
    }
    if (counts.values.contains(3)) {
      triples++;
    }
  }
  return doubles * triples;
}

dynamic part2(List<String> input) {
  for (var word1 in input) {
    for (var word2 in input) {
      var diffCount = 0;
      var common = '';
      assert(word1.length == word2.length);
      for (var k = 0; k < word1.length; k++) {
        if (word1[k] != word2[k]) {
          diffCount++;
        } else {
          common += word1[k];
        }
      }
      if (diffCount == 1) {
        return common;
      }
    }
  }
  return null;
}

main() {
  var input = readFile('input.txt');
  var testInput = readFile('test_input.txt');
  var testInput2 = readFile('test_input_2.txt');

  check(12, part1(testInput));
  print(part1(input));

  check("fgij", part2(testInput2));
  print(part2(input));
}
